src/OrleansRuntime/GrainDirectory/LocalGrainDirectory.cs

```diff
@@ -469,175 +472,235 @@ public SiloAddress CalculateTargetSilo(GrainId grain, bool excludeThisSiloIfStop
+                await Task.Delay(RETRY_DELAY);
+                forwardAddress = this.CheckIfShouldForward(address.Grain, hopCount, "RegisterAsync(recheck)");
             }
-            if (owner.Equals(MyAddress))
+
+            if (forwardAddress == null)
             {
-                RegistrationsLocal.Increment();
-                // if I am the owner, store the new activation locally
-                DirectoryPartition.AddActivation(address.Grain, address.Activation, address.Silo);
+                (singleActivation ? RegistrationsSingleActLocal : RegistrationsLocal).Increment();
+
+                // we are the owner     
+                var registrar = RegistrarManager.Instance.GetRegistrarForGrain(address.Grain);
+
+                return await registrar.RegisterAsync(address, singleActivation);
             }
             else
             {
-                RegistrationsRemoteSent.Increment();
+                (singleActivation ? RegistrationsSingleActRemoteSent : RegistrationsRemoteSent).Increment();
+
                 // otherwise, notify the owner
-                int eTag = await GetDirectoryReference(owner).Register(address, NUM_RETRIES);
-                if (IsValidSilo(address.Silo))
+                AddressAndTag result = await GetDirectoryReference(forwardAddress).RegisterAsync(address, singleActivation, hopCount + 1);
+
+                if (singleActivation)
                 {
-                    // Caching optimization:
-                    // cache the result of a successfull RegisterActivation call, only if it is not a duplicate activation.
+                    // Caching optimization: 
+                    // cache the result of a successfull RegisterSingleActivation call, only if it is not a duplicate activation.
                     // this way next local lookup will find this ActivationAddress in the cache and we will save a full lookup!
-                    IReadOnlyList<Tuple<SiloAddress, ActivationId>> cached;
-                    if (!DirectoryCache.LookUp(address.Grain, out cached))
+                    if (result.Address == null) return result;
+
+                    if (!address.Equals(result.Address) || !IsValidSilo(address.Silo)) return result;
+
+                    var cached = new List<Tuple<SiloAddress, ActivationId>>(1) { Tuple.Create(address.Silo, address.Activation) };
+                    // update the cache so next local lookup will find this ActivationAddress in the cache and we will save full lookup.
+                    DirectoryCache.AddOrUpdate(address.Grain, cached, result.VersionTag);
+                }
+                else
+                {
+                    if (IsValidSilo(address.Silo))
                     {
-                        cached = new List<Tuple<SiloAddress, ActivationId>>(1)
+                        // Caching optimization:
+                        // cache the result of a successfull RegisterActivation call, only if it is not a duplicate activation.
```

Comment:
I am not changing this for now as I do not fully understand it myself. I am just leaving it the way it was originally.